cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -854,12 +856,16 @@ void notifyListeners(List<FlowNode> nodes, boolean synchronous) {
     }
 
     @Restricted(DoNotUse.class)
-    @Terminator public static void suspendAll() throws InterruptedException, ExecutionException {
+    @Terminator public static void suspendAll() throws InterruptedException, ExecutionException, TimeoutException {
         LOGGER.fine("starting to suspend all executions");
         for (FlowExecution execution : FlowExecutionList.get()) {
             if (execution instanceof CpsFlowExecution) {
                 LOGGER.log(Level.FINE, "waiting to suspend {0}", execution);
-                ((CpsFlowExecution) execution).waitForSuspension();
+                CpsFlowExecution exec = (CpsFlowExecution) execution;
+                // Like waitForSuspension but with a timeout:
+                if (exec.programPromise != null) {
+                    exec.programPromise.get(1, TimeUnit.MINUTES).scheduleRun().get(1, TimeUnit.MINUTES);
+                }
```

Comment:
Should have fixed [JENKINS-28221](https://issues.jenkins-ci.org/browse/JENKINS-28221).